wgengine/magicsock/relaymanager.go

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package magicsock
+
+import (
+	"net/netip"
+	"sync"
+
+	"tailscale.com/disco"
+	"tailscale.com/types/key"
+)
+
+// relayManager manages allocation and handshaking of
+// [tailscale.com/net/udprelay.Server] endpoints.
+type relayManager struct {
+	mu                     sync.Mutex // guards the following fields
+	discoInfoByServerDisco map[key.DiscoPublic]*discoInfo
+}
+
+func newRelayManager() relayManager {
+	return relayManager{
+		discoInfoByServerDisco: make(map[key.DiscoPublic]*discoInfo),
+	}
+}
+
+// discoInfo returns a [*discoInfo] for 'serverDisco' if there is an
+// active/ongoing handshake with it, otherwise it returns nil, false.
+func (h *relayManager) discoInfo(serverDisco key.DiscoPublic) (*discoInfo, bool) {
```

Comment:
style nit: `(_ *discoInfo, ok bool)` ... try to always name "ok" bools

Reply:
updated in https://github.com/tailscale/tailscale/pull/15864/commits/5490636446733e876fedda8dd363d1e4d62630d1